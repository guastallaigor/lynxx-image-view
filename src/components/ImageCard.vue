<template>
  <div
    class="image-card"
    @click="goToFullScreenImage"
    :class="{ 'min-h-auto': minHeightAuto, 'min-h-300': !minHeightAuto }"
  >
    <v-lazy-image :src="src" :alt="caption" :title="caption" class="pointer" />
    <figcaption>{{ caption }}</figcaption>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from "vue-property-decorator";
import VLazyImage from "v-lazy-image";

@Component({
  components: {
    VLazyImage
  }
})
export default class ImageCard extends Vue {
  @Prop({ type: String, required: true })
  private id!: string;

  @Prop({ type: String, required: true })
  private src!: string;

  @Prop({ type: String, required: true })
  private caption!: string;

  @Prop({ type: Boolean, default: false })
  private minHeightAuto!: boolean;

  private goToFullScreenImage() {
    if (!this.id) return;

    this.$router.push(`/image/${this.id}`);
  }
}
</script>

<style lang="scss">
.v-lazy-image {
  filter: blur(10px);
  transition: filter 0.7s;
}
.v-lazy-image-loaded {
  filter: blur(0);
}
</style>

<style lang="scss" scoped>
.image-card {
  overflow: hidden;
  border-radius: 4px;
  width: 100%;
  min-height: 450px;
  transition: all 0.3s ease-in-out;
  will-change: opacity, top;
  position: relative;
  top: 0;

  &:hover {
    filter: opacity(80%);
    top: -5px;
  }

  img {
    border-radius: 4px;
    object-fit: cover;
    width: 100%;
    height: 100%;
    line-height: 0;
    display: block;
  }

  figcaption {
    background-color: $dark-gray-3;
    color: $white;
    font: italic smaller sans-serif;
    padding: 7px 5px 5px 5px;
    margin-top: -2px;
    text-align: center;
    border-bottom-right-radius: 4px;
    border-bottom-left-radius: 4px;
  }
}

.min-h-auto {
  @include lg-and-down-alternative() {
    min-height: auto;
  }
}

.min-h-300 {
  @include lg-and-down-alternative() {
    min-height: 300px;
  }
}
</style>
