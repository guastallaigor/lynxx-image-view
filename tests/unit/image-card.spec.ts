import { createLocalVue, shallowMount } from "@vue/test-utils";
import VueRouter from "vue-router";
import VLazyImage from "v-lazy-image";

import ImageCard from "@/components/ImageCard.vue";

describe("When I create the ImageCard component", () => {
  const createImageCardWrapper = (propsData = {}) => {
    const localVue = createLocalVue();
    localVue.use(VLazyImage);
    localVue.use(VueRouter);
    const router = new VueRouter();

    return shallowMount(ImageCard, {
      propsData,
      localVue,
      router
    });
  };

  const props = {
    id: "1",
    src: "https://ik.imagekit.io/6xhf1gnexgdgk/lion_BllLvaqVn.jpg",
    caption: "Lion",
    minHeightAuto: false
  };

  it(`should exists`, () => {
    const wrapper = createImageCardWrapper({ ...props });
    const imageCard = wrapper.find(".image-card");
    expect(wrapper.exists()).toBe(true);
    expect(imageCard.exists()).toBe(true);
    expect(imageCard.attributes().class).toContain("min-h-300");
  });

  it("should validate all props", () => {
    const consoleLog = console.error;
    console.error = jest.fn();
    const wrapper = createImageCardWrapper({ ...props });
    const { id }: any = wrapper?.vm?.$options?.props;
    expect(id.required).toBeTruthy();
    expect(id.type).toBe(String);
    expect(id.default).toBeUndefined();
    const { src }: any = wrapper?.vm?.$options?.props;
    expect(src.required).toBeTruthy();
    expect(src.type).toBe(String);
    expect(src.default).toBeUndefined();
    const { caption }: any = wrapper?.vm?.$options?.props;
    expect(caption.required).toBeTruthy();
    expect(caption.type).toBe(String);
    expect(caption.default).toBeUndefined();
    const { minHeightAuto }: any = wrapper?.vm?.$options?.props;
    expect(minHeightAuto.required).toBeFalsy();
    expect(minHeightAuto.type).toBe(Boolean);
    expect(minHeightAuto.default).toBeFalsy();
    console.error = consoleLog;
  });

  it(`should have an image`, () => {
    const wrapper = createImageCardWrapper({ ...props });
    expect(wrapper.exists()).toBe(true);
    const img = wrapper.find(".image-card>:nth-child(1)");
    expect(img.exists()).toBe(true);
    expect(img.attributes().class).toContain("pointer");
    expect(img.attributes().alt).toBe(props.caption);
    expect(img.attributes().title).toBe(props.caption);
    expect(img.attributes().src).toBe(props.src);
  });

  it(`should have a caption below the image`, () => {
    const wrapper = createImageCardWrapper({ ...props });
    expect(wrapper.exists()).toBe(true);
    const caption = wrapper.find(".image-card>:nth-child(2)");
    expect(caption.exists()).toBe(true);
    expect(caption.text()).toBe(props.caption);
  });

  it(`should change the minHeight to auto`, () => {
    const wrapper = createImageCardWrapper({ ...props, minHeightAuto: true });
    expect(wrapper.exists()).toBe(true);
    const imageCard = wrapper.find(".image-card");
    expect(imageCard.exists()).toBe(true);
    expect(imageCard.attributes().class).toContain("min-h-auto");
  });

  it(`should click and redirect to another URL`, async () => {
    const wrapper = createImageCardWrapper({ ...props });
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.vm.$route.path).toBe("/");
    const imageCard = wrapper.find(".image-card");
    expect(imageCard.exists()).toBe(true);
    imageCard.trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$route.path).toBe("/image/1");
    await wrapper.vm.$router.push("/");
  });

  it(`should block the click if has no id`, async () => {
    const consoleLog = console.error;
    console.error = jest.fn();
    const wrapper = createImageCardWrapper({ ...props, id: null });
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.vm.$route.path).toBe("/");
    const imageCard = wrapper.find(".image-card");
    expect(imageCard.exists()).toBe(true);
    imageCard.trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$route.path).toBe("/");
    console.error = consoleLog;
  });

  it("should match snapshot", () => {
    const wrapper = createImageCardWrapper({ ...props });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
