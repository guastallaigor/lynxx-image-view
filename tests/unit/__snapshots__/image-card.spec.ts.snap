// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the ImageCard component should match snapshot 1`] = `
<div class="image-card min-h-300">
  <v-lazy-image-stub src="https://ik.imagekit.io/6xhf1gnexgdgk/lion_BllLvaqVn.jpg" srcplaceholder="//:0" intersectionoptions="[object Object]" alt="Lion" title="Lion" class="pointer"></v-lazy-image-stub>
  <figcaption>Lion</figcaption>
</div>
`;
